tools/common-code/cmonitor_watcher.py

```diff
@@ -0,0 +1,196 @@
+#
+# Author: Satyabrata Bharati
+# Created: April 2022
+#
+import inotify.adapters
+import queue
+import os
+import time
+import logging
+from datetime import datetime
+
+exit_flag = False
+# =======================================================================================================
+# CgroupWatcher : Basic inotify class
+# =======================================================================================================
+class CgroupWatcher:
+    """
+    - Watch all files below a directory and notify an event for changes.
+    - Retrieves all the process and extract the process name "/proc/<pid>/stat.
+    - check the process name against the white-list given in the filter list.
+    - store the events in Queue.
+    """
+
+    def __init__(self, path, filter, timeout):
+        """Initialize CgroupWatcher
+        Args:
+          path: path to watch for events.
+          filter: white-list against which the process-event is filtered.
+
+        """
+        self.path = path
+        self.filter = filter
+        self.timeout = timeout
+        self.myFileList = {}
+
+    def __get_cgroup_version(self):
+        """
+        Detect the cgroup version.
+        """
+        proc_self_mount = "/proc/self/mounts"
+        ncgroup_v1 = 0
+        ncgroup_v2 = 0
+        with open(proc_self_mount) as file:
+            for line in file:
+                row = line.split()
+                fs_spec = row[0]
+                fs_file = row[1]
+                fs_vfstype = row[2]
+                if (fs_spec == "cgroup" or fs_spec == "cgroup2") and fs_vfstype == "cgroup2":
+                    ncgroup_v2 += 1
+                else:
+                    ncgroup_v1 += 1
+
+        if ncgroup_v1 == 0 and ncgroup_v2 > 0:
+            cgroup_versopn = "v2"
+            return cgroup_version
+        else:
+            cgroup_version = "v1"
+            return cgroup_version
+
+    def __get_process_name(self, pid):
+        """Returns the process name for the process id.
+        Args:
+          pid: process id.
+
+        Returns:
+          The process name.
+
+        """
+        cgroup_version = self.__get_cgroup_version()
+        if cgroup_version == "v1":
+            proc_filename = "/proc" + "/" + pid + "/stat"
+        else:
+            proc_filename = "/proc" + "/" + pid + "/cgroup.procs"
+        with open(proc_filename) as file:
+            for line in file:
+                parts = line.split()
+                process_name = parts[1].strip("()")
+        return process_name
+
+    def __get_pid_list(self, filename):
+        """Get the list of the process ids belong to a tasks file.
+        Args:
+          filename: the tasks file.
+
+        Returns:
+          The list of PIDs within the tasks file.
+
+        """
+        list = []
+        with open(filename) as file:
+            for line in file:
+                list.append(line.strip())
+        return list
+
+    def __get_list_of_files(self, dir):
```

Comment:
rename to __get_files_recursively
and remove mention of "event" from the description... this function doesn't deal with any event by itself... also remove "watched dir"... this function is generic